lib/ui/testing_hooks.dart

```diff
@@ -0,0 +1,19 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of dart.ui;
+
+/// @nodoc
+/// Makes some internal functions visible for internal Flutter testing.
+///
+/// Do not use for non-testing purposes, or catastrophic problems may (and
+/// probably will) occur.  These APIs are not part of the Flutter API, and are
+/// subject to change.  Do not depend upon their existence.
+class TestingHooks {
+  /// @nodoc
+  /// Make updateTextScaleFactor hook visible for testing.
+  /// Do not use for non-testing purposes: catastrophic problems may occur.
+  static void updateTextScaleFactor(double textScaleFactor) =>
+      _updateTextScaleFactor(textScaleFactor);
+}
```

Comment:
We should definitely not expose an API we don't want people to use.

Reply:
Any API we expose _will_ be used, and as soon as its used we can't break it, regardless of what our documentation says (especially if we hide it with `@nodoc`).

Reply:
Well, in principle, I agree, but there isn't any other way that I know of to simulate these events.  I discussed this with Jason before implementing it, and he seemed to think this was the best compromise.

Reply:
Out of curiosity: why are we _especially_ bound to @nodoc APIs?  Seems like just the opposite to me.

Reply:
We talked about this at lunch but for anyone watching:

It's probably not literally true that `@nodoc` is more likely to make people depend on it. There's a combination of factors that could push that way though:
- if we say `@nodoc`, then anything we put in the docs will be missed by anyone who decides to use it.
- people especially like to use secret APIs because it makes them feel cool and hip.
- if we say it's not supported then that lowers our expectation that people will use it, but it doesn't lower the actual probability that people will use it, so the end result is that relative to our expectation, people are more likely to end up depending on it.

At the end of the day though Flutter's convention is to not have "secret" APIs, we either have public documented supported APIs, or we don't expose the API at all, and in particular, our next layer should always use the previous layer's public API so that people can replace any layer and do the same thing we did (this is true at the macro level, e.g. "rendering" vs "dart:ui", as well as the micro level, e.g. Icon using IconTheme).

We want to avoid magic.